### ISSUE TITLE
Having trouble on calculating gradient of the network output w.r.t. input

### ISSUE THREAD

**Issue by JiamingSuen** [mit-han-lab/torchsparse#6]
Hello Haotian, thanks for this amazing library!

I'm currently working on a project with a need to calculate the gradient of the network output with respect to the network input explicitly. However, `torch.autograd` failed to do so and complaining `mm(): functions with out=... arguments don't support automatic differentiation, but one of the arguments requires grad.` I can't find anywhere in the library that calls `torch.mm(out=...)`. Do you have any idea how this issue is caused? The following snippet can be used to reproduce this error.

```python
import torch
import torchsparse
import torch.nn as nn
import torchsparse.nn as spnn
import numpy as np
from torchsparse import SparseTensor
from torchsparse.utils import sparse_quantize, sparse_collate_fn


def generate_random_point_cloud(size=100000, voxel_size=0.2):
    pc = np.random.randn(size, 4)
    pc[:, :3] = pc[:, :3] * 10
    rounded_pc = np.round(pc[:, :3] / voxel_size).astype(np.int32)
    labels = np.random.choice(10, size)
    inds, _, inverse_map = sparse_quantize(
        rounded_pc,
        pc,
        labels,
        return_index=True,
        return_invs=True
    )
    
    voxel_pc = rounded_pc[inds]
    voxel_feat = pc[inds]
    voxel_labels = labels[inds]
    
    sparse_tensor = SparseTensor(voxel_feat, voxel_pc)
    label_tensor = SparseTensor(voxel_labels, voxel_pc)
    
    feed_dict = {
        'lidar': sparse_tensor,
        'targets': label_tensor
    }
    
    return feed_dict


def generate_batched_random_point_clouds(
    size=10000, 
    voxel_size=0.2,
    batch_size=2):
    batch = []
    for i in range(batch_size):
        batch.append(generate_random_point_cloud(size, 
                                                 voxel_size))
    return sparse_collate_fn(batch)

device = 'cuda'

model = nn.Sequential(
    spnn.Conv3d(4, 32, kernel_size=3, stride=1),
    spnn.BatchNorm(32),
    spnn.ReLU(True),
).to(device)

feed_dict = generate_batched_random_point_clouds()
inputs = feed_dict['lidar'].to(device)
targets = feed_dict['targets'].F.to(device).long()
outputs = model(inputs)

coords = inputs.C.float()
sdf = outputs.F

coords = torch.autograd.Variable(coords, requires_grad=True)
grad = torch.autograd.grad(outputs=sdf, inputs=coords, 
                           grad_outputs=torch.ones(sdf.size(), device=sdf.device), 
                           create_graph=True, retain_graph=True, only_inputs=True)

```

Thanks a lot for your time!

**Comment by kentangSJTU** [comment]
Hi Jiaming,

Thanks for the question and sorry for the delayed response. I was preparing for an exam in the past two days. Regarding the error message from PyTorch, our library did use the `torch::mm` with `out` tensor specified [here](https://github.com/mit-han-lab/torchsparse/blob/f029da05ca7ef484783355f6e54507ad88b39a37/torchsparse/src/convolution/convolution.cpp#L72). I think this might partially explain the reason why you got this error.

For the code snippet you provided, I think it might be hard for you to get the gradient on the coordinates since they should not be in our computation graph. However, I do think getting the gradient for the input **features** is possible, if you run the following code:

```python
model.train()

feed_dict = generate_batched_random_point_clouds()
inputs = feed_dict['lidar'].to(device)
inputs.F.requires_grad = True
targets = feed_dict['targets'].F.to(device).long()
outputs = model(inputs)

loss = outputs.F.sum()
loss.backward()
print(inputs.F.grad)
```

after the model definition (i.e. the line containing `to(device)`). Here I define `loss = outputs.F.sum()` to achieve `grad_outputs = torch.ones(sdf.size(), device=sdf.device)`. Hope the code can help in your applications.


Best,
Haotian

**Comment by JiamingSuen** [comment]
Thanks for your reply!
I found that acquiring gradients with `autograd.grad` (instead of `loss.backward()`) requires `torch.mm` without `out=... arguments`. I attempted to change all `torch::mm_out()` calls in the C++ implementation to `torch::mm()` but the output of the network changed for some unknown reason. Since I've figured out another method to calculate the gradient without autograd, I will close this issue for now. Thanks for your help!
